### DISCUSSION TITLE
Upsampling from global representation

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#2786] [Q&A] [answered]
Hi Matt, I would like to ask about up-sampling from global pooled representations. 

Let say I have `encoder` and `decoder` like this to reconstruct original `input`.  
`SplineConv`->`cluster=graclus`->`max_pool`->`global_max(or mean)_pool`->`bottle neck`->`up-sampled[cluster]`->`SplineConv`.
When I have `global_mean_pool`, it takes average for each `out_channel` so we can get small dimensional `bottle neck` space like outputs from fully connected representations. 
However, as you said, when I tried to use `global_mean_pool`, each node holds the same feature value, which makes it difficult to obtain meaningful `up-sampled` values during decoding. 
If I must use `global_mean_pool`, (I need to have 1-dimensional representations at the `bottle neck` to train these representations with other dataset.), how would you `up-sample` from the `global_mean_pool` layer?

**Comment by rusty1s** [comment, marked as answer]
I think your adjacency matrix was fixed for each example, right? In this case this simplifies things a lot. You have two options:
1. Don't make use of global pooling and simply reshape your encoder output to `view(batch_size, -1)`, and reshape in decoder to `.view(num_coarsened_nodes * batch_size, -1)` again.
2. Make use of global pooling but make the first transformation in the decoder have different parameters for each node it maps to:
```python
lin = Linear(out_channels, num_coarsened_nodes * out_channels)
x = lin(z).view(num_coarsened_nodes * batch_size, -1)
```

**Reply by hkim716** [reply]
Thanks Matt. I have chosen second option because I needed to make the `bottle neck` smaller rather than just reshaping. 
My current code is this below. 

``` python
class autoencoder(torch.nn.Module):
    def __init__(self, in_channel, out_channel, kernel_size, batch_size):
        super(autoencoder, self).__init__()
        self.in_channel = in_channel
        self.out_channel = out_channel
        self.kernel_size = kernel_size
        self.batch_size = batch_size

        self.down_path = torch.nn.ModuleList()
        self.down_path.append(SplineConv(in_channel, out_channel, dim=2, kernel_size=kernel_size))
        self.down_path.append(SplineConv(out_channel, out_channel, dim=2, kernel_size=kernel_size))
        
        self.up_path = torch.nn.ModuleList()
        self.up_path.append(SplineConv(out_channel, out_channel, dim=2, kernel_size=kernel_size))
        self.up_path.append(SplineConv(out_channel, in_channel, dim=2, kernel_size=kernel_size))
        self.count = 0
        self.clusters = []
        self.latent_num_nodes = []

        self.lin = torch.nn.Linear(self.out_channel*self.batch_size, self.latent_num_nodes[-1]*self.out_channel)

      
    def forward(self, data):
        self.count += 1
        x, edge_index, edge_attr, pos = data.x.to(torch.float32), data.edge_index, data.edge_attr.to(torch.float32), data.pos.to(torch.float32)
        
        residuals = [] # intermediate node features : intermediate data.x
        edge_indices = []
        edge_attrs = []

        for i in range(len(self.down_path)):
        
            x = self.down_path[i](x, edge_index, edge_attr)
            
            residuals.append(x)
            
            edge_indices.append(edge_index)
            edge_attrs.append(edge_attr)
            
            if self.count == 1: # Let's fix the clustering indicies!
                cluster = graclus(edge_index)
                self.clusters.append(cluster)

            data.x = x

            data = max_pool(self.clusters[i], data, transform=transform)

            x, edge_index, edge_attr = data.x.to(torch.float32), data.edge_index, data.edge_attr.to(torch.float32)
            if self.count == 1:
                self.latent_num_nodes.append(data.num_nodes)

        # Global Representation
        global_x = global_max_pool(x, data.batch).view(-1)     
        
        x = self.lin(global_x)

        x = x.view(data.num_nodes, self.out_channel)

    
        for i in range(len(self.up_path)):
            cluster, _ = consecutive_cluster(self.clusters[-i-1]) # -1, -2, -3, ...
            
            res = residuals[-i-1]
            edge_index = edge_indices[-i-1]
            edge_attr = edge_attrs[-i-1]
            
            upsampled_x = x[cluster]

            x = self.up_path[i](upsampled_x, edge_index, edge_attr)

        return x.view(-1).to(torch.float32)
```

I tried to put `linear` layer after `global_pooling` layer, but it was difficult to define 
`self.lin = torch.nn.Linear(self.out_channel*self.batch_size, self.latent_num_nodes[-1]*self.out_channel)`
because I should have `self.latent_num_nodes[-1]` (which means the number of nodes of the most coarsened data) with `data.num_nodes` at `self.count=1` (which means forward pass for the first batch.)

I tried to update `self.latent_num_node` with the `num_nodes` from the latest `max_pool` when computing the first batch, like this,
``` python
            if self.count == 1:
                self.latent_num_nodes.append(data.num_nodes)
```
but `self.lin = torch.nn.Linear(self.out_channel*self.batch_size, self.latent_num_nodes[-1]*self.out_channel)` here, `self.latent_num_nodes` was still defined as empty list `[ ]` at the first forward pass, which cause error.  

How could you define `self.lin`, updating `self.latent_num_nodes=data.num_nodes` from the `max_pool` for the first forward pass?

**Reply by rusty1s** [reply]
Yes, the number of coarsened nodes needs to be known in advance. You can probably get it by running `torch.unique(dataset[0].cluster).numel()` before model instantiation.

**Reply by hkim716** [reply]
Thanks Matt! I was able to define `linear` layer using `torch.unique(cluster).numel()` before the model instantiation. 
Now I'm trying to implement the first option because the results from the second option does not look good. (I think using `global_pooling` layer at `bottle neck` reduces a lot of performance because of the small latent space. ) 

I have question about the second option you mentioned.  You said I could reshape encoder output to `view(batch_size, -1)`, and reshape in decoder to `.view(num_coarsened_nodes * batch_size, -1)` again.
However, the output of `x.shape` is `[2708, 16]`, which is `[clusters[-1].shape, out_channel]`, and because my `batch_size=100`, I cannot reshape output to `view(100, -1)`. Moreover, I'm also curious how to see `num_coarsened_nodes`.
I know that `graclus` clusters nodes randomly, and I guess the number of clusters, `2708`, means the number of clusters for each `batch`. Am I wrong at some point? How could you reshape the outputs from `max_pool` like this?

``` python
data = max_pool(self.clusters[-1], data, transform=transform)
x = data.x.reshape(batch_size, -1)
```

**Reply by rusty1s** [reply]
The second method does only work if your `cluster` vector is fixed for each example. Given that you have a decoder output of `[2708, 16]` and a `batch_size` of 100, the number of clusters should be `2708 // 100`. However, you are right that this does not give you an integer value back (but it should :D). Aren't you using the same `cluster` vector for each graph?

**Reply by hkim716** [reply]
You are right, and I meant the output does not gave me an integer value, `2798//100`. I think `2708//100` is not very meaningful value if it's not an integer.  
Yes, I'm using the same `cluster` vector for each graph. Then in that case, could I obtain `cluster` vector like this?

``` python
def pre_process(data, batch_size, depth):
    clusters = []
    for i in range(depth):
        edge_index=data.edge_index
        
        stopper = False
        while stopper == False:
            if i != depth-1:
                cluster = graclus(edge_index)
                break
            else :    
                cluster = graclus(edge_index)
                if cluster.size(0) % batch_size == 0:
                    stopper = True
        clusters.append(cluster)
        data = max_pool(clusters[i], data, transform=transform)
            
    return clusters
```
In this case, it finds the last cluster that satisfy the condition `cluster.size(0) % batch_size == 0` with small `batch_sizes` like `2`, `4`, etc., but it didn't with large values of `batch_size > 10`. Can we actually make `graclus` search broader range of number of clusters?

**Reply by rusty1s** [reply]
I'm not sure I understand. I wouldn't pre-process the `cluster` vector in a batch-wise fashion, since it will produce different clusters for examples in a batch. I think your issue from above is that there might be an issue of batching the `cluster` vector. Did you modify the `Data.__inc__` method correctly for this attribute?

**Reply by hkim716** [reply]
Understood. You are right. I tried to generate `cluster` vector using an example graph and store the `cluster` to `data` object. However, it caused me an error when using large `batch_size`. 
So I decided not to store the `cluster` vector into the `data` object, and I tried to use the list of `clusters[i]` obtained from the `pre_process` function, which does not use `Data.__inc__`. 
Is there another way to produce same `cluster` vector for examples in a batch without storing `cluster` vector to `data` object?

**Reply by rusty1s** [reply]
You can also hold it globally (as you currently do), and then do:
```python
max_clusters = int(cluster.max())
batched_cluster = torch.cat([cluster] * batch_size) + data.batch * max_clusters
```
This will correctly create the `cluster` vector in a mini-batch scenario.

**Reply by hkim716** [reply]
I tried to make `batched_cluster`. However, the shape of `data.batch * max_clusters` does not match with the size of `torch.cat([cluster] * batch_size)` for the second and third coarsened `clusters`. (In the case of `depth=3`)

``` python
def pre_process(train_dataset, depth):
    train_loader_pre = DataLoader(train_dataset, batch_size=1, shuffle=False)
    clusters = []
    data_batches = []
    for data in train_loader_pre:
        for i in range(depth):
            x, edge_index, edge_attr, pos = data.x.to(torch.float32), data.edge_index, data.edge_attr.to(torch.float32), data.pos.to(torch.float32)
            cluster = graclus(edge_index)
            clusters.append(cluster)
            data = max_pool(clusters[i], data, transform=transform)
        break
    return clusters

def batched_cluster(data, batch_size, clusters, depth):
    batched_clusters = []
    for i in range(depth):
        max_clusters = int(clusters[i].max())
        batched_cluster = torch.cat([clusters[i]] * batch_size).to(device) + data.batch * max_clusters
        batched_clusters.append(batched_cluster)
    return batched_clusters

clusters = pre_process(train_dataset, 3)

batch_size = 5
train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=False)
test_loader = DataLoader(test_dataset, batch_size=batch_size)


for data in train_loader:
    data = data.to(device)
    batched_clusters = batched_cluster(data, batch_size, clusters, depth=3)
    break
```

`batched_clusters[0]` should work because `torch.cat([clusters[0]] * batch_size)` and `data.batch` have the same shape, but how could you generate `batched_clusters[1], [2], etc` when there's no information of the second coarsened `data.batch`?

**Reply by rusty1s** [reply]
Two things are missing here:
1. You need to run `consecutive cluster` after computing `cluster`.
2. You need to modify `data.batch` in each level of `batched_cluster`, e.g.:
```python
batch = data.batch
for i in range(depth):
    cluster = clusters[i]
    max_clusters = int(cluster.max())
    batched_cluster = torch.cat([cluster] * batch_size).to(device) + batch * max_clusters
    batch = torch.repeat_interleave(torch.arange(batch_size), cluster.size(0))
```

**Reply by hkim716** [reply]
Thanks Matt, it worked, and I think `batched_clusters` are successfully generated.
However, there was another shape mismatch issue during `forward` function. (the mismatch did not happen when I tried with `pre-process` in a batch-wise fashion)
A part of my `forward` function looks like this. 
``` python
def forward(self, data):
    x, edge_index, edge_attr = data.x, data.edge_index, data.edge_attr

    for i in range(len(self.down_path)):
        print(x.shape)
        x = self.down_path[i](x, edge_index, edge_attr)
        data.x = x
        print(self.batched_clusters[i].shape)
        data = max_pool(self.batched_clusters[i], data, transform=transform)
        print(data.x.shape)
        x, edge_index, edge_attr = data.x, data.edge_index, data.edge_attr
```
When I checked these three shapes, the results are

```
torch.Size([370, 32])
torch.Size([370])
torch.Size([206, 32])

torch.Size([206, 32])
torch.Size([210])
```

I think `Size([206, 32])` should be `Size([210, 32])`, which is `([batched_clusters[1], hidden_channels])`. I think `206` came from `max_pool` operation, which does not care about the next `batched_cluster[1]`. How could I make this shape same with the next `batched_cluster`?

The shapes of three `batched_clusters` generated here are 
```
torch.Size([370])
torch.Size([210])
torch.Size([130])
```
and I found the difference of `210` and `206` was always `210-206=4`, which was `batch_size-1`. 
When I use `batch_size=100`, the difference was `99`. It's weird..

**Reply by hkim716** [reply]
Oh, I found what the problem was!
`batched_cluster = torch.cat([cluster] * batch_size).to(device) + batch * max_clusters`
should have been `batched_cluster = torch.cat([cluster] * batch_size).to(device) + batch * (max_clusters+1)`.

**Reply by rusty1s** [reply]
Oh, you are right. Thanks for fixing it :)
